tech-specs/src/explore/design/dkg.md

```diff
@@ -1 +1,154 @@
-# Distributed key generation gossip
+# Distributed key generation
+
+Here we give an overview of how the DKG protocol works with the ledger and 
+the roles of each node in the network per this protocol. This portion of the 
+book focuses solely on the production of a new public encryption key via the
+DKG scheme. It does not cover encryption, decryption, or validation of transactions.
+
+Some parts of the DKG protocol may be done through the gossip layer or through
+off-chain communication. However, the entire DKG cannot exist solely in the 
+gossip layer as consensus on the final shared secrets must be achieved. This 
+will require posting information on chain.
+
+## Starting the Protocol
+
+To initializing the protocol to generate a new key, the  current block 
+proposer proposes a new instance. This would be posted on the blockchain and
+notify all the participating validators to begin the announcement phase. 
+However, when should the proposer make this decision? In all likelihood it will be a combination of blocks transpired since
+the last instance completed as well as special cases (such as an instance
+failing to complete).
+
+## Announcing and distributing weight shares
+
+In order for the DKG protocol to work, it requires the contribution of 
+ \\(\frac{2}{3} \\) of the total staked weight among live validators. In 
+order to know when a subset of validators own at least \\(\frac{2}{3}\\)
+of the total weight, it is necessary for validators to announce their stake.
+
+Furthermore, the total weight will be scaled down to a fixed parameter 
+\\( W\\) (and fractional weights rounded appropriately to acheive integral
+values). After this procedure, each validator has a corresponding number of
+weight shares. 
+
+Lastly, the DKG protocol requires a canonical ordering of validators with
+shares must be agreed upon. This will be done via the consensus layer.
+The current block proposer will apportion out weight shares as well as the
+aforementioned ordering (which induces a natural partition of the weight 
+shares) to each validator that made an announcement. This is included in the
+block. Once this block is on the blockchain, the next step of the DKG can proceed.
+
+## Dealers
+In a single DKG instance, a subset of validators holding at least
+\\( \frac{2}{3} \\) of the weight shares must each act as dealers. It
+is the responsibility of a dealer to post a signed message to the blockchain
+which includes the session id (\\( \tau \\)) as well the data of the
+PVSS (we do not concern ourselves with the details here). The creation of this
+message is handled by the Ferveo library.
+
+This means that the signed message is a transaction that must be gossiped and
+eventually included on the blockchain. For this particular transaction, there
+is a special verification that must be done, otherwise the transaction should
```

Comment:
This is going to be a little messy: we want to do signature verification on the "deal" transaction, but otherwise defer all verification until a later stage. The reason for this is that in the "optimistic" case all of the PVSS messages can be batch-verified. This will save a lot of compute, but we'll have to add a complaint message as well to handle the pessimistic case.

Reply:
So this part is a bit out of date. We will get the `Deal` messages as vote extensions, not as transactions. So all of the verification can happen in `VerifyVoteExtension` in bulk. Since that doesn't trigger until 2/3 validators have responded, the optimistic case should be good.  If something goes wrong, complaints  (instead of an aggregation) can be added to header instead I suppose.

Reply:
Hmm, yeah. We want the complaint case (either a bad `Deal` or bad `Aggregate`) to be punished, so hopefully it never happens (and certainly not many times in a row) but we do have to handle it somehow in a not-too-expensive way (I am a bit concerned that generating the `Complaint` might be rather slow)

Reply:
I'm realizing that I haven't spent much time thinking about punishments at all. I will add a section to the spec about it but it is probably worth some brainstorming as to how we want to achieve it.